File: client_compatibility/python/runner.py
import json
import os
import sys

try:
    import pymysql
except ImportError:
    print("PyMySQL is not installed", file=sys.stderr)
    sys.exit(125)

dsn = os.environ["XMYSQL_CLIENT_DSN"]
host, port = dsn.split(":", 1)
connection = pymysql.connect(host=host, port=int(port), user=os.environ["XMYSQL_CLIENT_USER"], password=os.environ["XMYSQL_CLIENT_PASSWORD"], autocommit=True, charset="utf8mb4")
try:
    with connection.cursor() as cursor:
        cursor.execute("SELECT 1")
        assert cursor.fetchone()[0] == 1
        cursor.execute("CREATE DATABASE IF NOT EXISTS client_matrix")
        cursor.execute("CREATE TABLE IF NOT EXISTS client_matrix.matrix_rows(id INT PRIMARY KEY, label VARCHAR(32))")
        cursor.execute("INSERT INTO client_matrix.matrix_rows(id, label) VALUES (1, %s) ON DUPLICATE KEY UPDATE label=VALUES(label)", ("one",))
        cursor.execute("SELECT %s + 1", (41,))
        assert cursor.fetchone()[0] == 42
        connection.begin()
        cursor.execute("INSERT INTO client_matrix.matrix_rows(id, label) VALUES (2, %s)", ("tx",))
        connection.rollback()
        cursor.execute("SELECT NULL, CAST(42 AS SIGNED), _utf8mb4'兼容'")
        row = cursor.fetchone()
        assert row == (None, 42, "兼容"), row
        cursor.execute("SELECT COUNT(*) FROM information_schema.COLUMNS WHERE TABLE_SCHEMA='client_matrix'")
        assert cursor.fetchone()[0] > 0
    print(json.dumps({"client": "pymysql", "status": "PASS"}, ensure_ascii=False))
finally:
    connection.close()

File: client_compatibility/node/runner.js
const mysql = (() => { try { return require('mysql2/promise'); } catch (_) { return null; } })();
if (!mysql) { console.error('mysql2 is not installed'); process.exit(125); }

(async () => {
  const connection = await mysql.createConnection({host: process.env.XMYSQL_CLIENT_HOST, port: Number(process.env.XMYSQL_CLIENT_PORT), user: process.env.XMYSQL_CLIENT_USER, password: process.env.XMYSQL_CLIENT_PASSWORD, charset: 'utf8mb4'});
  try {
    const [one] = await connection.query('SELECT 1');
    if (one[0]['1'] !== 1) throw new Error('SELECT 1 failed');
    await connection.query('CREATE DATABASE IF NOT EXISTS client_matrix');
    await connection.query('CREATE TABLE IF NOT EXISTS client_matrix.matrix_rows(id INT PRIMARY KEY, label VARCHAR(32))');
    await connection.execute('INSERT INTO client_matrix.matrix_rows(id, label) VALUES (?, ?) ON DUPLICATE KEY UPDATE label=VALUES(label)', [1, 'one']);
    const [prepared] = await connection.execute('SELECT ? + 1 AS value', [41]);
    if (prepared[0].value !== 42) throw new Error('prepared statement failed');
    await connection.beginTransaction();
    await connection.execute('INSERT INTO client_matrix.matrix_rows(id, label) VALUES (?, ?)', [2, 'tx']);
    await connection.rollback();
    const [typed] = await connection.query("SELECT NULL AS nullable_value, CAST(42 AS SIGNED) AS integer_value, _utf8mb4'兼容' AS utf8mb4_value");
    if (typed[0].nullable_value !== null || typed[0].integer_value !== 42 || typed[0].utf8mb4_value !== '兼容') throw new Error('NULL/type/charset failed');
    const [metadata] = await connection.query("SELECT COUNT(*) AS count FROM information_schema.COLUMNS WHERE TABLE_SCHEMA='client_matrix'");
    if (metadata[0].count <= 0) throw new Error('metadata failed');
    await connection.end();
    console.log(JSON.stringify({client: 'node-mysql2', status: 'PASS'}));
  } catch (error) { console.error(error.stack || error); process.exit(1); }
})();
